cibuildwheel/projectfiles.py

```diff
@@ -0,0 +1,82 @@
+import ast
+import sys
+from configparser import ConfigParser
+from pathlib import Path
+from typing import Any, Dict, Optional
+
+import toml
+
+if sys.version_info < (3, 8):
+    Constant = ast.Str
+
+    def get_constant(x: ast.Str) -> str:
+        return x.s
+else:
+    Constant = ast.Constant
+
+    def get_constant(x: ast.Constant) -> Any:
+        return x.value
+
+
+class Analyzer(ast.NodeVisitor):
+    def __init__(self) -> None:
+        self.requires_python: Optional[str] = None
+        self.constants: Dict[str, str] = {}
+
+    def visit_Assign(self, node: ast.Assign) -> None:
+        for target in node.targets:
+            if (
+                isinstance(target, ast.Name)
+                and isinstance(node.value, Constant)
+                and isinstance(get_constant(node.value), str)
+            ):
+                self.constants[target.id] = get_constant(node.value)
+
+    def visit_keyword(self, node: ast.keyword) -> None:
+        self.generic_visit(node)
+        if node.arg == "python_requires":
+            if isinstance(node.value, Constant):
+                self.requires_python = get_constant(node.value)
+            elif isinstance(node.value, ast.Name):
+                self.requires_python = self.constants.get(node.value.id)
+
+
+def setup_py_python_requires(content: str) -> Optional[str]:
+    try:
+        tree = ast.parse(content)
+        analyzer = Analyzer()
```

Comment:
This is very cool, but I'm wondering if it's worth the code. All these projects are probably already setting `CIBW_SKIP`, anyway?

Reply:
For the record, you've done all the work already and I don't think this is something that would be maintained often, so we can just as well keep it? :-)

Reply:
The goal is to improve the first run experience, that is, #468 - this was just a useful database to run from. You shouldn't have to set up build selectors just to limit your Python versions to the versions you support. A _lot_ of projects use static metadata in setup.py still (mostly due to old examples, old tutorials, and thinking that using "newer" techniques will cause it to break for some users - though it's almost [always the reverse](https://github.com/scikit-hep/uproot4/pull/231), from my experience, and there's a good chance that dependencies use setup.cfg for metadata already, like twine and wheel). It would be nice if, when adding cibuildwheel to a new project, it would just behave nicely, and this improves the chances of that happening.

I actually originally wrote the logic for a script to help with moving from setup.py to setup.cfg, and just adapted it here.

Reply:
Selecting the things to be built has so far always worked, without any complaints. To me, `CIBW_BUILD` and `CIBW_SKIP` are pretty basic options, so I don't fully agree that this is a must-have for the first run experience.
It's a nice-to-have feature, yes, but still not really bullet-proof, either, so I'm not entirely convinced this is worth the code. But not against keeping it, either.

Reply:
Pretty sure I linked to at least one complaint in my previous comment. ;)

Reply:
Hmmm, valid point :-D I thought I had remembered that title to be an internal discussion, but obviously not, if you take the time to actually read.

Reply:
It was renamed, originally was a bug report for Python 2 showing up in 2020. :)

Reply:
And arguably not wrong, as bug report, without a bit of context ;-)

Reply:
This is even cooler: https://github.com/henryiii/setuptools-modernize/blob/main/src/setuptools_modernize/parse.py :)